pkg/serror/serror.go

```diff
@@ -0,0 +1,173 @@
+)
+
+// Well-known keys for structured error attributes
+const (
+	// MessageKey is the key used for the main error message.
+	// The associated value is a string.
+	MessageKey = "msg"
+
+	// CauseKey is the key used for the underlying wrapped error.
+	// The associated value is an error.
+	CauseKey = "cause"
+
+	// SourceKey is the key used for for the source file and line of the error site.
+	// The associated value is a *[Source].
+	SourceKey = "source"
+
+	// ErrorCodeKey is the key used for the status code associated with the error.
+	// In the context of HTTP, this is the status code. In the context of
+	// an application, this is an application-specific error code.
+	// The associated value is an int.
+	ErrorCodeKey = "code"
+
+	// TimeKey is the key used for the time when the error was created.
+	// The associated value is a [time.Time].
+	TimeKey = "time"
+)
+
+// Error represents a structured error with attributes that conform to slog conventions.
+// Error is immutable - methods like With() return new instances rather than
+// modifying the original error.
+type Error struct {
+	record slog.Record
+	cause  error
+}
+
+// Ensure Error implements the error interface
+var _ error = Error{}
+
+// Ensure Error implements slog.LogValuer
+var _ slog.LogValuer = Error{}
+
+// For testing
+var timeNow = time.Now
+
+// New creates a new Error with the given message and attributes.
+// The attributes should be alternating key-value pairs, similar to slog conventions.
+func New(msg string, args ...any) Error {
+	// Capture caller's program counter
+	var pcs [1]uintptr
+	runtime.Callers(2, pcs[:]) // skip runtime.Callers + New
+
+	e := new(timeNow(), msg, nil, pcs[0])
+	e.add(args...)
+
+	return e
+}
+
+func new(t time.Time, msg string, cause error, pc uintptr) Error {
+	return Error{
+		cause:  cause,
+		record: slog.NewRecord(t, slog.LevelError, msg, pc),
+	}
+}
+
+// Wrap creates a new Error that wraps an existing error with additional context.
+// If cause is nil, it behaves like New.
+func Wrap(cause error, msg string, args ...any) Error {
+	// Capture caller's program counter
+	var pcs [1]uintptr
+	runtime.Callers(2, pcs[:]) // skip runtime.Callers + Wrap
+
+	e := new(timeNow(), msg, cause, pcs[0])
+	e.add(args...)
+
+	return e
+}
+
+func (e *Error) add(args ...any) {
+	if e == nil {
+		return
+	}
+	e.record.Add(args...)
+}
+
+func (e Error) Error() string {
+	var buf bytes.Buffer
+	e.format(&buf)
+	return strings.TrimSuffix(buf.String(), "\n")
+}
+
+func (e Error) format(w io.Writer) {
+	// Create a new record with all the information
+	record := e.record.Clone()
+	record.Time = time.Time{} // Set time to zero so we don't print it
+
+	// If we have a cause, add it as an error attribute
+	if e.cause != nil {
+		record.AddAttrs(slog.String("cause", e.cause.Error()))
+	}
+
+	// Create a tint handler that writes to our writer
+	handler := tint.NewHandler(w, &tint.Options{
+		Level:   slog.LevelError,
+		NoColor: true,
```

Comment:
wait, why use tint if we are not using color?

Reply:
Did you mean for something like https://github.com/lmittmann/tint?tab=readme-ov-file#automatically-enable-colors ?